components/affix/affix.directive.ts

```diff
@@ -0,0 +1,174 @@
+import {Directive, OnInit, Input, Output, EventEmitter, ElementRef, OnDestroy, HostBinding} from 'angular2/core';
+import {positionService, ElemPosition} from '../position';
+
+export enum AffixStatus {AFFIX, AFFIX_TOP, AFFIX_BOTTOM}
+
+export class AffixStatusChange {
+    constructor(public oldStatus:AffixStatus, public newStatus:AffixStatus) {
+    }
+}
+
+@Directive({
+    selector: '[affix]'
+})
+export class Affix implements OnInit, OnDestroy {
+
+    @Input()
+    public affixOffsetTop:number = 0;
+    @Input()
+    public affixOffsetBottom:number = 0;
+
+    @HostBinding('class.affix')
+    private isAffix:boolean = true;
+    @HostBinding('class.affix-top')
+    private isAffixedTop:boolean = true;
+    @HostBinding('class.affix-bottom')
+    private isAffixedBottom:boolean = true;
+    @HostBinding('style.top.px')
+    private top:number = null;
+
+    @Output()
+    public affixChange:EventEmitter<AffixStatusChange> = new EventEmitter(false);
+
+    private status:AffixStatus = null;
+    private body:HTMLBodyElement;
+    private window:Window;
+    private pinnedOffset:number = null;
+    private debouncedCheckPosition:Function = Affix.debounce(() => this.checkPosition(), 5);
+    private eventListener:Function = (ev:UIEvent) => this.debouncedCheckPosition();
+
+    constructor(private el:ElementRef) {
+        this.body = el.nativeElement.ownerDocument.body;
+        this.window = el.nativeElement.ownerDocument.defaultView;
+    }
+
+    ngOnInit() {
+        this.el.nativeElement.ownerDocument.defaultView.addEventListener('scroll', this.eventListener);
+        this.checkPosition();
+    }
+
+    ngOnDestroy():any {
+        this.el.nativeElement.ownerDocument.defaultView.removeEventListener('scroll', this.eventListener);
+        return undefined;
+    }
+
+    private checkPosition():void {
+        let elemPos = positionService.position(this.el.nativeElement);
```

Comment:
This call can actually fail, though not due to problems in this pull request.  See https://github.com/valor-software/ngx-bootstrap/pull/2042